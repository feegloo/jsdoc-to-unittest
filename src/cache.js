import { parse } from 'acorn';
import { isPrimitive } from './utils';

const acornDefaults = {
  ecmaVersion: 8,
};

const generic = (cache = new WeakMap()) => (target, onMiss) => {
  if (isPrimitive(target)) {
    return onMiss(target);
  }

  if (cache.has(target)) {
    return cache.get(target);
  }

  const value = onMiss(target);
  cache.set(target, value);
  return value;
};

export const acorn = (() => {
  const cache = generic();
  return {
    parse(target, setup = {}) {
      return cache(target, () => {
        try {
          return parse(target, { ...acornDefaults, ...setup });
        } catch (ex) {
          return parse(`(${target})`, { ...acornDefaults, ...setup });
        }
      });
    },
  };
})();

export default generic();
