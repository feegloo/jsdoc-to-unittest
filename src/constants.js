export default {
  jest: {
    method: 'test',
  },

  karma: {
    method: 'it',
  },
};

export const tokens = {
  CALL: 'call',
  ACCESS: 'access',
  GET: 'get',
};
