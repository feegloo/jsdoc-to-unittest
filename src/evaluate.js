import { getFunctionBody } from './utils';

export default (src, _eval) => _eval(getFunctionBody(src));

