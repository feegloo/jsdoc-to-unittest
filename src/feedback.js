import { isAsyncFunction } from './utils';

const knownObjects = new Set();

const feedback = (value) => {
  if (arguments.length === 0) {
    return ['generic'];
  }

  const type = [typeof value, 'generic'];

  switch (type[0]) {
    case 'function':
      if (isAsyncFunction(value)) {
        type.unshift('AsyncFunction');
      }
      break;
    case 'object':
      if (value === null) {
        type[0] = 'null';
      } else if (Array.isArray(value)) {
        if (knownObjects.has(value)) {
          return value;
        }

        type.unshift('Array');
      } else if (value instanceof Date) {
        type.unshift('Date');
      }
      break;
    default:
  }

  knownObjects.add(type);
  return type;
};

export function matchesFeedback(received, expected) {

}

export const Types = {
  Any: [],
  Primitive: ['primitive'],
  Boolean: feedback(true),
  Number: feedback(2),
  String: feedback(''),
  Symbol: feedback(Symbol.iterator),
  Null: feedback(null),
  AsyncFunction: feedback(async () => {}),
  Function: feedback(() => {}),
  Date: feedback(new Date()),
  Object: feedback({}),
  Array: feedback([]),
  'Array<Smi>': feedback([0]),
};

Object.values(Types).forEach(knownObjects.add, knownObjects);

export default feedback;
