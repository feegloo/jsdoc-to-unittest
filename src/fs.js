import fs from 'fs';
import util from 'util';

export const writeFileAsync = util.promisify(fs.writeFile);
export const readFileAsync = util.promisify(fs.readFile);

export const toStdout = (output) => {
  if (process.env.NODE_ENV === 'test') {
    return output;
  }

  process.stdout.write(output);
  return true;
};
