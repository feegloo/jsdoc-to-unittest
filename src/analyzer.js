import { evaluate, evaluateAsync, getFunctionBody, getFunctionParams, isPrimitive } from './utils';
import feedback, { Types } from './feedback';
import joinPath from './path';
import mockedConsole from '../__mocks__/console';

function intercept(ret, _eval, args = {}, _async = false) {
  const obj = {
    path: [],
    access: [],
    calls: 0,
    mapped: {},
  };
  const seenKeys = new Map();
  let current = null;
  let currentKey = '';
  let included = false;
  let excluded = false;
  const instance = new Proxy(function () {}, { // eslint-disable-line prefer-arrow-callback
    apply(target, thisArg, argumentsList) {
      if (argumentsList.some(item => item === instance)) {
        ret.pop();
        current = ret[ret.length - 1];
      }

      current.calls += 1;
      const argsList = new Array(argumentsList.length);
      for (let i = 0; i < argumentsList.length; i += 1) {
        argsList[i] = feedback(argumentsList[i]);
      }

      const intercepted = ['call', argsList];
      current.access.push(...intercepted);
      if (currentKey !== undefined && current.mapped[currentKey] !== undefined) {
        current.mapped[currentKey].push(...intercepted);
      }

      try {
        const ref = _eval(joinPath(current.path));
        if (typeof ref === 'function') {
          const parsedArgs = {};
          getFunctionParams(ref).forEach((name, i) => {
            parsedArgs[name] = argumentsList[i];
          });
          ret.push(...parseKey(getFunctionBody(ref), _eval, parsedArgs));
        }
      } catch (ex) {}

      return instance;
    },
    construct(target, argumentsList) {
      if (argumentsList.some(item => item === instance)) {
        ret.pop();
        current = ret[ret.length - 1];
      }
      return instance;
    },
    has(target, key) {
      if (!Reflect.has(args, key)) {
        current = {
          ...JSON.parse(JSON.stringify(obj)),
          mapped: {
            * [Symbol.iterator]() {
              const mappedKeys = Object.keys(this);
              while (mappedKeys.length) {
                const mappedKey = mappedKeys.shift();
                yield [
                  mappedKey.replace(/\$\d+/, ''),
                  this[mappedKey],
                ];
              }
            },
          },
        };
        ret.push(current);
        seenKeys.clear();
        included = false;
        excluded = false;
      } else {
        excluded = true;
      }
      return true;
    },
    get(target, key) {
      switch (key) {
        case Symbol.unscopables:
          return void 0; // eslint-disable-line no-void
        case Symbol.toPrimitive:
          // todo: move it outside the closure,
          // as there is no need to allocate this function over and over again...
          return function (hint) {
            target.apply(this, arguments); // eslint-disable-line prefer-rest-params
            switch (hint) {
              case 'number':
                return 0;
              case 'string':
                return '';
              default:
                return true;
            }
          };
        case 'toJSON':
          return () => null; // todo as above
        default:
          if (key === '_asyncTestExports') {
            break;
          }

          if (!excluded) {
            current.path.push(key);
            currentKey = key;
            let index = seenKeys.get(currentKey);
            if (index) {
              currentKey = `${currentKey}$${index}`;
            } else {
              index = 0;
            }

            if (!included) {
              included = true;
              current.mapped[currentKey] = ['access'];
              current.access.push('access');
            } else {
              current.mapped[currentKey] = ['get'];
              current.access.push('get');
            }

            seenKeys.set(currentKey, index + 1);
          }
      }

      if (Reflect.has(args, key)) {
        return args[key];
      }

      if (_async && current.access.includes('get') && /then(\$\d*)?/.test(currentKey)) {
        return;
      }

      return instance;
    },
  });

  return instance;
}

export function parseKey(exp, _eval = () => {}, args = {}) {
  const ret = [];
  evaluate(`with(s){\n${exp}\n}`, {
    s: intercept(ret, _eval, { ...args, Types }),
  });
  return ret;
}

export async function parseKeyAsync(exp, _eval = () => {}, args = {}) {
  const ret = [];
  await evaluateAsync(`with(s){\n${exp}\n}`, {
    s: intercept(ret, _eval, {
      Promise,
      Types,
      ...args,
    }, true),
  });
  return ret;
}

export function getPath(func) {
  return parseKey(func, this).reduce((acc, { path }) => [...acc, path], []);
}

export const getPathAsync = async (func) => {
  const parsed = await parseKeyAsync(func, this);
  return parsed.reduce((acc, { path }) => [...acc, path], []);
};

export function isCallable(func, filter, all = false) {
  return parseKey(func, this)
    .filter(({ path }) => filter.includes(path.join('.')))
    [all ? 'every' : 'some'](({ calls }) => calls > 0); // eslint-disable-line no-unexpected-multiline
}

function sandbox(func) {
  return Function(
    's, ...args',
    `s.args=args;with(s){return(${func.toString()}).apply(this,arguments)}`,
  );
}

export function fallToGlobal(target, _eval, oldKey = '') {
  const mappedArgs = new WeakMap();
  const privateScope = new WeakMap();
  privateScope.set(target, {});
  return new Proxy(target, {
    has: () => true,
    set(_, key, value) {
      if (key === 'args') {
        mappedArgs.set(target, value);
      } else {
        privateScope.get(target)[key] = value;
      }

      return true;
    },
    get(_, key) {
      if (key === '_asyncTestExports') { // fixme: that _asyncTestExports
        return fallToGlobal(target, _eval);
      }

      if (key === Symbol.unscopables) return null;
      if (key === 'arguments') {
        return mappedArgs.get(target) || [];
      }

      const scope = privateScope.get(target);
      if (key in scope) {
        return scope[key];
      }

      if (!isPrimitive(target) && Reflect.has(target, key)) {
        const newTarget = target[key];
        if (typeof newTarget !== 'object' && typeof newTarget !== 'function') {
          return newTarget;
        }

        return fallToGlobal(newTarget, _eval, `${oldKey}.${key}`);
      }

      try {
        let func;
        if (oldKey) {
          func = _eval(`${oldKey}.${key}`.replace('.', ''));
        } else {
          func = _eval(key);
        }

        if (typeof func === 'function') {
          return sandbox(func).bind(null, fallToGlobal(target, _eval, `${oldKey}.${key})`));
        }

        return func;
      } catch (ex) {}
    },
  });
}

function filterAccesses(parsed, filter) {
  if (filter.length) {
    const paths = filter.reduce((acc, path) => {
      acc.push(...parseKey(path));
      return acc;
    }, []);

    parsed = parsed.map((obj) => {
      const { mapped } = obj;
      const keyedMapped = Object.keys(mapped);
      let max = -1;
      let foundObj;
      paths.forEach(({ mapped: filterMapped }) => {
        Object.keys(filterMapped).every((key, i) => {
          if (key === keyedMapped[i]) {
            if (i > max) {
              max = i;
              foundObj = obj;
            }
            return true;
          }


          return false;
        });
      });


      return foundObj;
    }).filter(item => item);
  }

  return parsed.reduce((acc, { access }) => [...acc, access], []);
}

export function listAccesses(code, filter = []) {
  try {
    return filterAccesses(parseKey(code, this), filter);
  } catch (ex) {
    return [];
  }
}

export async function listAccessesAsync(code, filter = []) {
  try {
    return filterAccesses(await parseKeyAsync(code, this), filter);
  } catch (ex) {
    return [];
  }
}

export async function useConsoleLog(code, filter) {
  const logs = [];
  await evaluateAsync(code, {
    console: mockedConsole.revokable(logs),
  });
  return logs.length > 0;
}
