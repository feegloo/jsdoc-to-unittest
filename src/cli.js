import path from 'path';
import yargs from 'yargs';
import { readFileAsync } from './fs';
import print from './creator';

yargs
  .usage('Usage: $0 <command> [options]')
  .demand('input')
  .default('target', 'jest')
  .default('inline', false)
  .boolean('inline');

const { argv } = yargs;

const basePath = path.resolve(__dirname, '..');

const fields = [
  'inline',
  'target',
];

const files = [
  'mocks',
  'intro',
  'header',
  'outro',
  'footer',
];

async function getConfig(input, output = '') {
  let configPath = '';
  const config = { // default values
    export: true,
    target: 'jest',
  };
  if (typeof argv.config === 'string') {
    try {
      JSON.parse(argv.config);
    } catch (ex) {
      configPath = path.resolve(basePath, argv.config, '..');
      Object.assign(config, JSON.parse(
        await readFileAsync(path.resolve(basePath, argv.config), 'utf-8'),
      ));
    }
  }

  fields.forEach((field) => {
    if (field in argv) {
      config[field] = argv[field];
    }
  });

  Object.assign(config, {
    input,
    output,
  });

  await Promise.all(files
    .filter(key => key in argv || key in config)
    .map(async (key) => {
      config[key] = await readFileAsync(
        path.resolve(argv[key] ? basePath : configPath, config[key]),
        'utf-8',
      );
    }),
  );

  return config;
}

export default async (input = argv.input, output = argv.output) => {
  try {
    const config = await getConfig(input, output);
    return print({
      ...config,
      get stdout() {
        return output === undefined || String(output).trim().length === 0;
      },
    });
  } catch (ex) {
    console.error(ex);
    return false;
  }
};
