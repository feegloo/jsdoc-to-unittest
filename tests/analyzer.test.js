import {
  parseKey,
  parseKeyAsync,
  fallToGlobal,
  getPath,
  isCallable,
  listAccesses,
  listAccessesAsync,
} from 'src/analyzer';
import { Types } from 'src/feedback';

const iterator = {
  * [Symbol.iterator]() {
    const mappedKeys = Object.keys(this);
    while (mappedKeys.length) {
      const mappedKey = mappedKeys.shift();
      yield [
        mappedKey.replace(/\$\d+/, ''),
        this[mappedKey],
      ];
    }
  },
};

describe('#parseKey', () => {
  test('direct', () => {
    expect(parseKey('foo')).toEqual([{
      calls: 0,
      access: ['access'],
      path: ['foo'],
      mapped: {
        foo: ['access'],
        ...iterator,
      },
    }]);

    expect(parseKey('foo()')).toEqual([{
      calls: 1,
      access: ['access', 'call', []],
      path: ['foo'],
      mapped: {
        foo: ['access', 'call', []],
        ...iterator,
      },
    }]);
  });

  test('getters', () => {
    expect(parseKey('foo.bar.baz')).toEqual([{
      calls: 0,
      access: ['access', 'get', 'get'],
      path: ['foo', 'bar', 'baz'],
      mapped: {
        foo: ['access'],
        bar: ['get'],
        baz: ['get'],
        ...iterator,
      },
    }]);

    expect(parseKey('foo[\'bar\'].baz')).toEqual([{
      calls: 0,
      access: ['access', 'get', 'get'],
      path: ['foo', 'bar', 'baz'],
      mapped: {
        foo: ['access'],
        bar: ['get'],
        baz: ['get'],
        ...iterator,
      },
    }]);
  });

  test('returns #mapped in correct order', () => {
    expect(Object.keys(parseKey('foo[\'bar\'].baz')[0].mapped)).toEqual(['foo', 'bar', 'baz']);
  });

  test('mapped doesn\'t contain the same keys', () => {
    expect(Object.keys(parseKey('foo.baz.baz')[0].mapped)).toEqual(['foo', 'baz', 'baz$1']);
  });

  xtest('generators', () => {
  })

  test('getters and calls', () => {
    expect(parseKey('foo.bar.baz(1, 2);foo.baz.baz.boo')).toEqual([
      {
        calls: 1,
        access: ['access', 'get', 'get', 'call', [Types.Number, Types.Number]],
        path: ['foo', 'bar', 'baz'],
        mapped: {
          foo: ['access'],
          bar: ['get'],
          baz: ['get', 'call', [Types.Number, Types.Number]],
          ...iterator,
        },
      },
      {
        calls: 0,
        access: ['access', 'get', 'get', 'get'],
        path: ['foo', 'baz', 'baz', 'boo'],
        mapped: {
          foo: ['access'],
          baz: ['get'],
          baz$1: ['get'],
          boo: ['get'],
          ...iterator,
        },
      },
    ]);
  });

  test('calls with primitives', () => {
    expect(parseKey('foo.bar.baz(1, 2)')).toEqual([{
      calls: 1,
      access: ['access', 'get', 'get', 'call', [Types.Number, Types.Number]],
      path: ['foo', 'bar', 'baz'],
      mapped: {
        foo: ['access'],
        bar: ['get'],
        baz: ['get', 'call', [Types.Number, Types.Number]],
        ...iterator,
      },
    }]);
  });

  test('nested access', () => {
    expect(parseKey('foo[0].bar().baz(1, 2)')).toEqual([{
      path: ['foo', '0', 'bar', 'baz'],
      calls: 2,
      access: ['access', 'get', 'get', 'call', [], 'get', 'call', [Types.Number, Types.Number]],
      mapped: {
        foo: ['access'],
        0: ['get'],
        bar: ['get', 'call', []],
        baz: ['get', 'call', [Types.Number, Types.Number]],
        ...iterator,
      },
    }]);

    expect(parseKey('foo[0].bar(10, 12).baz(1, 2);baz.bar(10, 12);')).toEqual([
      {
        calls: 2,
        path: ['foo', '0', 'bar', 'baz'],
        access: [
          'access',
          'get',
          'get',
          'call',
          [Types.Number, Types.Number],
          'get',
          'call',
          [Types.Number, Types.Number],
        ],
        mapped: {
          foo: ['access'],
          0: ['get'],
          bar: ['get', 'call', [Types.Number, Types.Number]],
          baz: ['get', 'call', [Types.Number, Types.Number]],
          ...iterator,
        },
      },
      {
        calls: 1,
        path: ['baz', 'bar'],
        access: ['access', 'get', 'call', [Types.Number, Types.Number]],
        mapped: {
          baz: ['access'],
          bar: ['get', 'call', [Types.Number, Types.Number]],
          ...iterator,
        },
      },
    ]);
  });

  test('nested functions', () => {
    function contains() {
      return forEach();
    }

    expect(parseKey('contains()', func => eval(func))).toEqual([
      {
        access: ['access', 'call', []],
        calls: 1,
        path: ['contains'],
        mapped: {
          contains: ['access', 'call', []],
          ...iterator,
        },
      },
      {
        access: ['access', 'call', []],
        calls: 1,
        path: ['forEach'],
        mapped: {
          forEach: ['access', 'call', []],
          ...iterator,
        },
      },
    ]);
  });

  test('return statement', () => {
    expect(parseKey('return foo.bar.baz(1, 2)')).toEqual([{
      calls: 1,
      access: ['access', 'get', 'get', 'call', [Types.Number, Types.Number]],
      path: ['foo', 'bar', 'baz'],
      mapped: {
        foo: ['access'],
        bar: ['get'],
        baz: ['get', 'call', [Types.Number, Types.Number]],
        ...iterator,
      },
    }]);

    expect(parseKey(`{
        return xD();
    }`)).toEqual([{
      calls: 1,
      access: ['access', 'call', []],
      path: ['xD'],
      mapped: {
        xD: ['access', 'call', []],
        ...iterator,
      },
    }]);
  });

  test('analyzes promises as expected', () => {
    expect(parseKey('obj.foo().then')).toEqual([{
      calls: 1,
      access: ['access', 'get', 'call', [], 'get'],
      path: ['obj', 'foo', 'then'],
      mapped: {
        obj: ['access'],
        foo: ['get', 'call', []],
        then: ['get'],
        ...iterator,
      },
    }]);

    expect(parseKey('obj.foo().then(() => {}).catch(() => {})')).toEqual([{
      calls: 3,
      access: ['access', 'get', 'call', [], 'get', 'call', [Types.Function], 'get', 'call', [Types.Function]],
      path: ['obj', 'foo', 'then', 'catch'],
      mapped: {
        obj: ['access'],
        foo: ['get', 'call', []],
        then: ['get', 'call', [Types.Function]],
        catch: ['get', 'call', [Types.Function]],
        ...iterator,
      },
    }]);
  });

  test('Symbol.unscopables', () => {
    expect(parseKey('forEach()')).toEqual([{
      calls: 1,
      access: ['access', 'call', []],
      path: ['forEach'],
      mapped: {
        forEach: ['access', 'call', []],
        ...iterator,
      },
    }]);
  });

  test('currying', () => {
    expect(parseKey('x()(true)(10)([2])')).toMatchSnapshot();
  })

  test('calls with non-primitives', () => {
    expect(parseKey('easy.utils.isDate(new Date()); // true')).toEqual([{
      calls: 1,
      access: ['access', 'get', 'get', 'call', [Types.Date]],
      path: ['easy', 'utils', 'isDate'],
      mapped: {
        easy: ['access'],
        utils: ['get'],
        isDate: ['get', 'call', [Types.Date]],
        ...iterator,
      },
    }]);

    expect(parseKey('easy.utils.isDate(easy); // true')).toEqual([{
      calls: 1,
      access: ['access', 'get', 'get', 'call', [{ type: 'object', value: null }]],
      path: ['easy', 'utils', 'isDate'],
      mapped: {
        easy: ['access'],
        utils: ['get'],
        isDate: ['get', 'call', [{ type: 'object', value: null }]],
        ...iterator,
      },
    }]);

    expect(parseKey('easy.utils.isDate(new Date(new Date())); // true')).toEqual([{
      calls: 1,
      access: ['access', 'get', 'get', 'call', [
        { type: 'object', value: null },
      ]],
      path: ['easy', 'utils', 'isDate'],
      mapped: {
        easy: ['access'],
        utils: ['get'],
        isDate: ['get', 'call', [{ type: 'object', value: null }]],
        ...iterator,
      },
    }]);
  });

  test('toPrimitive', () => {
    expect(parseKey('foo.bar() + xyz()')).toEqual([
      {
        access: ['access', 'get', 'call', []],
        calls: 1,
        path: ['foo', 'bar'],
        mapped: {
          foo: ['access'],
          bar: ['get', 'call', []],
          ...iterator,
        },
      },
      {
        access: ['access', 'call', []],
        calls: 1,
        path: ['xyz'],
        mapped: {
          xyz: ['access', 'call', []],
          ...iterator,
        },
      }]);

    expect(parseKey('foo.bar() + +xyz()')).toEqual([
      {
        access: ['access', 'get', 'call', []],
        calls: 1,
        path: ['foo', 'bar'],
        mapped: {
          foo: ['access'],
          bar: ['get', 'call', []],
          ...iterator,
        },
      },
      {
        access: ['access', 'call', []],
        calls: 1,
        path: ['xyz'],
        mapped: {
          xyz: ['access', 'call', []],
          ...iterator,
        },
      }]);

    expect(() => parseKey('foo.bar[Symbol.hasPrimitive]()')).not.toThrow();
  });

  test('raises SyntaxError', () => {
    expect(() => parseKey(';;;-/\\\\--;;;')).toThrow(SyntaxError);
  });

  test('mapped properties can be iterated', () => {
    const { mapped } = parseKey('foo.foo.baz.boo.baz.bar')[0];
    expect(Object.keys(mapped)).toEqual(['foo', 'foo$1', 'baz', 'boo', 'baz$1', 'bar']);
    expect([...mapped]).toEqual([
      ['foo', ['access']],
      ['foo', ['get']],
      ['baz', ['get']],
      ['boo', ['get']],
      ['baz', ['get']],
      ['bar', ['get']],
    ]);
  });

  test('collects valid feedback', () => {
    function getFromURL(str) {
      return fetch(str); // eslint-disable-line no-undef
    }

    expect(parseKey('getFromURL(\'x\')', func => eval(func)))
      .toEqual([
        {
          path: ['getFromURL'],
          access: ['access', 'call', [Types.String]],
          calls: 1,
          mapped: {
            getFromURL: ['access', 'call', [Types.String]],
            ...iterator,
          },
        },
        {
          path: ['fetch'],
          access: ['access', 'call', [Types.String]],
          calls: 1,
          mapped: {
            fetch: ['access', 'call', [Types.String]],
            ...iterator,
          },
        },
      ]);
  });
});

describe('#parseKeyAsync', () => {
  test('analyzes async functions', async () => {
    expect(await parseKeyAsync(`{
      foo.call();
      await someAsyncCall(20);
      foo.await();
    }`)).toMatchSnapshot();
  });
});

describe('#listAccesses', () => {
  test('lists accesses', () => {
    expect(listAccesses('frosmo', ['frosmo'])).toEqual([['access']]);
    expect(listAccesses('frosmo()', ['frosmo'])).toEqual([['access', 'call', []]]);
    expect(listAccesses('frosmo.xyz.aaa', ['frosmo.xyz.aaa'])).toEqual([['access', 'get', 'get']]);
    expect(listAccesses('frosmo.xyz["aaa"]', ['frosmo.xyz.aaa'])).toEqual([['access', 'get', 'get']]);
    expect(listAccesses('frosmo.xyz.aaa', ['frosmo.xyz["aaa"]'])).toEqual([['access', 'get', 'get']]);
    expect(listAccesses('frosmo.xyz.aaa()', ['frosmo.xyz.aaa'])).toEqual([['access', 'get', 'get', 'call', []]]);
    expect(listAccesses('frosmo.xyz.aaa(true, false, 2)', ['frosmo.xyz.aaa()']))
      .toEqual([['access', 'get', 'get', 'call', [Types.Boolean, Types.Boolean, Types.Number]]]);
    expect(listAccesses('frosmo.xyz.aaa();frosmo.xyz.aaa', ['frosmo.xyz.aaa'])).toEqual([
      ['access', 'get', 'get', 'call', []],
      ['access', 'get', 'get'],
    ]);
  });

  test('detects correct number of arguments', () => {
    expect(listAccesses('frosmo.xyz().aaa(true, false, 2)', ['frosmo.xyz.aaa()'])).toEqual(
      [['access', 'get', 'call', [], 'get', 'call', [Types.Boolean, Types.Boolean, Types.Number]]],
    );
  });

  test('has optional filter', () => {
    expect(listAccesses('lol.xx();frosmo.xyz.aaa(true, false, 2)', [])).toEqual([
      ['access', 'get', 'call', []],
      ['access', 'get', 'get', 'call', [Types.Boolean, Types.Boolean, Types.Number]],
    ]);

    expect(listAccesses('lol.xx();frosmo.xyz.aaa(true, false, 2)'))
      .toEqual(listAccesses('lol.xx();frosmo.xyz.aaa(true, false, 2)', []));
  });

  test('shadows objects', () => {
    expect(listAccesses('lol.xx();frosmo.xyz.aaa(true, false, 2)', ['frosmo.xyz.aaa()'])).toEqual([
      ['access', 'get', 'get', 'call', [Types.Boolean, Types.Boolean, Types.Number]]
    ]);
  });

  test('filters partial chains', () => {
    expect(listAccesses('foo.bar.a;foo.bar.c.x', ['foo.bar'])).toEqual([
      ['access', 'get', 'get'],
      ['access', 'get', 'get', 'get'],
    ]);

    expect(listAccesses('foo.bar.a;foo.bar.c.x', ['foo.bar', 'foo.bar.a'])).toEqual([
      ['access', 'get', 'get'],
      ['access', 'get', 'get', 'get'],
    ]);
  });

  test('nested calls', () => {
    function contains(str) {
      return foo.bar.baz(str);
    }

    expect(listAccesses.call(func => eval(func), 'contains(2)')).toEqual([
      ['access', 'call', [Types.Number]],
      ['access', 'get', 'get', 'call', [Types.Number]],
    ]);
  });

  test('never throws', () => {
    expect(listAccesses('baz.foo.bar()', ['frosmo.xyz.aaa()'])).toEqual([]);
    expect(listAccesses('baz.fo;;---o.bar()', ['frosmo.xyz.aaa()'])).toEqual([]);
  });
});

describe('#listAccessAsync', () => {
  test('lists accesses', async () => {
    expect(await listAccessesAsync(`
      const response = await getFromURL('https://mail.google.com/mail/');
      return response.slice(5).trim()`,
    ))
      .toEqual([
        ['access', 'call', [Types.String]],
        ['access', 'get', 'call', [Types.Number], 'call', []],
      ]);
  });

  test('never throws', async () => {
    expect(await listAccessesAsync('baz.foo.bar()', ['frosmo.xyz.aaa()'])).toEqual([]);
    expect(await listAccessesAsync('baz.fo;;---o.bar()', ['frosmo.xyz.aaa()'])).toEqual([]);
  });
});

describe('#getPath', () => {
  test('get paths correctly', () => {
    expect(getPath('frosmo.test.xyz.call()')).toEqual([['frosmo', 'test', 'xyz', 'call']]);
    expect(getPath('frosmo.yes.hello.call')).toEqual([['frosmo', 'yes', 'hello', 'call']]);
    expect(getPath('easy.easy.hello.call')).toEqual([['easy', 'easy', 'hello', 'call']]);
  });

  test('follows the flow execution', () => {
    expect(getPath('false && frosmo.yes.hello.call')).toEqual([]);
    expect(getPath('false && frosmo.test.xyz.call()')).toEqual([]);
  });

  test('brackets notated', () => {
    expect(getPath('easy[\'bar\'][\'baz\']')).toEqual([['easy', 'bar', 'baz']]);
    expect(getPath('frosmo[0][1][\'bar\']')).toEqual([['frosmo', '0', '1', 'bar']]);
  });

  test('multiple chains', () => {
    expect(getPath('frosmo.test.xyz.call();easy.abc.foo.bar'))
      .toEqual([['frosmo', 'test', 'xyz', 'call'], ['easy', 'abc', 'foo', 'bar']]);
  });

  test('returns first key if no path is present', () => {
    expect(getPath('random')).toEqual([['random']]);
  });

  test('throws SyntaxError', () => {
    expect(() => getPath('---')).toThrow(SyntaxError);
  });

  test('simple nested calls', () => {
    function contains() {
      return foo.bar.baz();
    }

    expect(getPath.call(func => eval(func), 'contains()', true)).toEqual([
      ['contains'],
      ['foo', 'bar', 'baz'],
    ]);
  });

  test('nested calls', () => {
    function forEach() {
      return foo.bar.baz();
    }

    function contains() {
      return forEach([1, 2]);
    }

    expect(getPath.call(func => eval(func), 'contains()', true)).toEqual([
      ['contains'],
      ['forEach'],
      ['foo', 'bar', 'baz'],
    ]);
  });

  test('nested calls #2', () => {
    function forEach(func) {
      return func();
    }

    function context() {
      return foo.bar.baz;
    }

    function contains() {
      return forEach(context());
    }

    expect(getPath.call(func => eval(func), 'contains()', true)).toEqual([
      ['contains'],
      ['forEach'],
      ['context'],
      ['foo', 'bar', 'baz'],
    ]);
  });

  test('\'inaccessible\' nested calls', () => {
    function forEach(func) {
      return func();
    }

    function contains() {
      function context() {
        return foo.bar.baz;
      }

      return forEach(context());
    }

    expect(getPath.call(func => eval(func), 'contains()', true)).toEqual([
      ['contains'],
      ['forEach'],
      ['context'],
      ['foo', 'bar', 'baz'],
    ]);
  });
});

describe('#isCallable', () => {
  test('works correctly', () => {
    expect(isCallable('easy.bar.yes()', ['easy.bar.yes'])).toBe(true);
    expect(isCallable('frosmo.yes()', ['frosmo.yes'])).toBe(true);
    expect(isCallable('frosmo.yes', ['frosmo.yes'])).toBe(false);
    expect(isCallable('frosmo.xyz()', ['frosmo.abc'])).toBe(false);
  });

  test('multiple chains', () => {
    expect(isCallable('foo.bar();foo.bar', ['foo.bar'])).toBe(true);
    expect(isCallable('foo.bar();foo.bar()', ['foo.bar'])).toBe(true);
  });

  test('multiple chains with all flag', () => {
    expect(isCallable('foo.bar();foo.bar', ['foo.bar'], true)).toBe(false);
    expect(isCallable('foo.bar();foo.bar()', ['foo.bar'], true)).toBe(true);
    expect(isCallable('foo[0].bar();foo.bar()', ['foo[0].bar'], true)).toBe(true);
  });
});

describe('#fallToGlobal', () => {
  test('gives upon SyntaxError', () => {
    expect(fallToGlobal({}, eval)['---;;;;---']).toBe(undefined);
  });
});
