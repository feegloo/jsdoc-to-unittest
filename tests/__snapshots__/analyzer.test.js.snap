// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parseKey currying 1`] = `
Array [
  Object {
    "access": Array [
      "access",
      "call",
      Array [],
      "call",
      Array [
        Array [
          "boolean",
          "generic",
        ],
      ],
      "call",
      Array [
        Array [
          "number",
          "generic",
        ],
      ],
      "call",
      Array [
        Array [
          "Array",
          "object",
          "generic",
        ],
      ],
    ],
    "calls": 4,
    "mapped": Object {
      "x": Array [
        "access",
        "call",
        Array [],
        "call",
        Array [
          Array [
            "boolean",
            "generic",
          ],
        ],
        "call",
        Array [
          Array [
            "number",
            "generic",
          ],
        ],
        "call",
        Array [
          Array [
            "Array",
            "object",
            "generic",
          ],
        ],
      ],
      Symbol(Symbol.iterator): [Function],
    },
    "path": Array [
      "x",
    ],
  },
]
`;

exports[`#parseKeyAsync analyzes async functions 1`] = `
Array [
  Object {
    "access": Array [
      "access",
      "get",
      "call",
      Array [],
    ],
    "calls": 1,
    "mapped": Object {
      "call": Array [
        "get",
        "call",
        Array [],
      ],
      "foo": Array [
        "access",
      ],
      Symbol(Symbol.iterator): [Function],
    },
    "path": Array [
      "foo",
      "call",
    ],
  },
  Object {
    "access": Array [
      "access",
      "call",
      Array [
        Array [
          "number",
          "generic",
        ],
      ],
      "get",
    ],
    "calls": 1,
    "mapped": Object {
      "someAsyncCall": Array [
        "access",
        "call",
        Array [
          Array [
            "number",
            "generic",
          ],
        ],
      ],
      "then": Array [
        "get",
      ],
      Symbol(Symbol.iterator): [Function],
    },
    "path": Array [
      "someAsyncCall",
      "then",
    ],
  },
  Object {
    "access": Array [
      "access",
      "get",
      "call",
      Array [],
    ],
    "calls": 1,
    "mapped": Object {
      "await": Array [
        "get",
        "call",
        Array [],
      ],
      "foo": Array [
        "access",
      ],
      Symbol(Symbol.iterator): [Function],
    },
    "path": Array [
      "foo",
      "await",
    ],
  },
]
`;
